File: app/src/main/java/org/secuso/privacyfriendlytodolist/tutorial/TutorialActivity.java
package org.secuso.privacyfriendlytodolist.tutorial;


import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.view.PagerAdapter;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.text.Html;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;

import org.secuso.privacyfriendlytodolist.R;
import org.secuso.privacyfriendlytodolist.view.MainActivity;

/**
 *
 * @author sebbi
 * @version 20171206
 *
 */

public class TutorialActivity extends AppCompatActivity{

    private PrefManager prefManager;
    public static final String TAG = TutorialActivity.class.getSimpleName();
    public static final String ACTION_SHOW_ANYWAYS = TAG + ".ACTION_SHOW_ANYWAYS";
    private ViewPager viewPager;
    private LinearLayout dotsLayout;
    private Button btnSkip, btnNext;
    private int[] layouts;
    private TextView[] dots;
    private MyViewPageAdapter myViewPageAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        //Check if app is startet the first time
        prefManager = new PrefManager(this);
        Intent i = getIntent();

        if (!prefManager.isFirstTimeLaunch() && (i == null || !ACTION_SHOW_ANYWAYS.equals(i.getAction()))){
            launchHomeScreen();
            return;
        }

        // Make notification bar transparent
        if(Build.VERSION.SDK_INT >= 21) {
            getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_STABLE | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN);
        }

        setContentView(R.layout.activity_tutorial);
        viewPager = (ViewPager)findViewById(R.id.view_pager);
        dotsLayout = (LinearLayout)findViewById(R.id.layoutDots);
        btnSkip = (Button)findViewById(R.id.btn_skip);
        btnNext = (Button)findViewById(R.id.btn_next);

        //add slides to layouts array
        layouts = new int[] {
                R.layout.tutorial_slide1,
                R.layout.tutorial_slide2,
                R.layout.tutorial_slide3
        };

        //add bottom dots
        addBottomDots(0);

        //change statusbar to transparent
        changeStatusBarColor();

        myViewPageAdapter = new MyViewPageAdapter();
        viewPager.setAdapter(myViewPageAdapter);
        viewPager.addOnPageChangeListener(viewPagerPageChangeListener);
        btnSkip.setOnClickListener(new View.OnClickListener(){
            @Override
            public void onClick (View v){
                launchHomeScreen();
            }
        });

        btnNext.setOnClickListener(new View.OnClickListener()
        {
            @Override
            public void onClick (View v){
                // checking for last page
                // if last page home screen will be launched
                int current = getItem(+1);
                if (current < layouts.length) {
                    // move to next screen
                    viewPager.setCurrentItem(current);
                } else {
                    launchHomeScreen();
                }
            }
        });
    }


    private void launchHomeScreen(){
        prefManager.setFirstTimeLaunch(false);
        Intent intent = new Intent(TutorialActivity.this, MainActivity.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
        startActivity(intent);
        finish();
    }

    ViewPager.OnPageChangeListener viewPagerPageChangeListener = new ViewPager.OnPageChangeListener() {
        @Override
        public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
        }

        @Override
        public void onPageSelected(int position) {
            addBottomDots(position);

            // change button text 'NEXT' on last slide to 'GOT IT'
            if (position == layouts.length - 1){
                // last slide
                btnNext.setText("Got it!");
                btnSkip.setVisibility(View.GONE);
            } else{
                // not last slide reached yet
                btnNext.setText("Next");
                btnSkip.setVisibility(View.VISIBLE);
            }
        }

        @Override
        public void onPageScrollStateChanged(int state) {
        }
    };


    private void addBottomDots(int currentPage){
        dots = new TextView[layouts.length];
        int[] colorsActive = getResources().getIntArray(R.array.array_dot_active);
        int[] colorsInactive = getResources().getIntArray(R.array.array_dot_inactive);

        dotsLayout.removeAllViews();
        for (int i=0; i < dots.length; i++){
            dots[i] = new TextView(this);
            dots[i].setText(Html.fromHtml("&#8226;"));
            dots[i].setTextSize(35);
            dots[i].setTextColor(colorsInactive[currentPage]);
            dotsLayout.addView(dots[i]);
        }

        if(dots.length > 0){
            dots[currentPage].setTextColor(colorsActive[currentPage]);
        }
    }


    private void changeStatusBarColor(){
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP){
            Window window = getWindow();
            window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
            window.setStatusBarColor(Color.TRANSPARENT);
        }
    }


    private int getItem(int i){
        return viewPager.getCurrentItem() + i;
    }




    private class MyViewPageAdapter extends PagerAdapter{
        private LayoutInflater layoutInflater;

        public MyViewPageAdapter(){
        }

        public Object instantiateItem(ViewGroup container, int position){
            layoutInflater = (LayoutInflater) getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            View view = layoutInflater.inflate(layouts[position], container, false);
            container.addView(view);
            return view;
        }

        @Override
        public int getCount() {
            return layouts.length;
        }

        @Override
        public boolean isViewFromObject(View view, Object object) {
            return view == object;
        }

        public void destroyItem(ViewGroup container, int position, Object object){
            View view = (View) object;
            container.removeView(view);
        }

    }


}

File: app/src/main/java/org/secuso/privacyfriendlytodolist/view/TutorialActivity.kt
/*
Privacy Friendly To-Do List
Copyright (C) 2017-2024  Sebastian Lutz

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/
package org.secuso.privacyfriendlytodolist.view

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.LinearLayout
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.viewpager.widget.PagerAdapter
import androidx.viewpager.widget.ViewPager
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import org.secuso.privacyfriendlytodolist.R
import org.secuso.privacyfriendlytodolist.util.PreferenceMgr

/**
 * Created by Sebastian Lutz on 06.12.2017.
 *
 * This Activity sets up the tutorial that shall appear for the first start of the app.
 */
class TutorialActivity : AppCompatActivity() {
    private lateinit var viewPager: ViewPager
    private val slides = arrayOf(
        R.layout.tutorial_slide1,
        R.layout.tutorial_slide2,
        R.layout.tutorial_slide3,
        R.layout.tutorial_slide4,
        R.layout.tutorial_slide5)

    private var colorActive = 0
    private var colorInactive = 0
    private val dots = mutableListOf<TextView>()
    private lateinit var dotsLayout: LinearLayout

    private lateinit var btnNext: Button
    private lateinit var btnSkip: Button

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // Make notification bar transparent
        setContentView(R.layout.activity_tutorial)
        viewPager = findViewById(R.id.view_pager)
        colorActive = ContextCompat.getColor(this, R.color.dotActive)
        colorInactive = ContextCompat.getColor(this, R.color.dotInactive)
        dotsLayout = findViewById(R.id.layoutDots)
        btnNext = findViewById(R.id.btn_next)
        btnSkip = findViewById(R.id.btn_skip)

        //add bottom dots
        addDots()

        viewPager.setAdapter(TutorialViewPageAdapter())
        viewPager.addOnPageChangeListener(viewPagerPageChangeListener)
        btnSkip.setOnClickListener {
            launchHomeScreen()
        }
        btnNext.setOnClickListener {
            // checking for last page
            // if last page home screen will be launched
            val nextItem = viewPager.currentItem + 1
            if (nextItem < slides.size) {
                // move to next screen
                viewPager.setCurrentItem(nextItem)
            } else {
                launchHomeScreen()
            }
        }
    }

    private fun launchHomeScreen() {
        PreferenceMgr.setFirstTimeLaunch(this, false)
        val intent = Intent(this@TutorialActivity, MainActivity::class.java)
        intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
        startActivity(intent)
        finish()
    }

    private var viewPagerPageChangeListener: OnPageChangeListener = object : OnPageChangeListener {
        override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
        }

        override fun onPageSelected(position: Int) {
            highlightDot(position)

            // change button text 'NEXT' on last slide to 'GOT IT'
            if (position == slides.size - 1) {
                // last slide
                btnNext.setText(R.string.okay)
                btnSkip.visibility = View.GONE
            } else {
                // not last slide reached yet
                btnNext.setText(R.string.next)
                btnSkip.visibility = View.VISIBLE
            }
        }

        override fun onPageScrollStateChanged(state: Int) {}
    }

    private fun addDots() {
        for (i in slides.indices) {
            val dot = TextView(this)
            dot.text = DOT
            dot.textSize = 35f
            dot.setTextColor(if (i == 0) colorActive else colorInactive)
            dots.add(dot)
            dotsLayout.addView(dot)
        }
    }

    private fun highlightDot(currentPage: Int) {
        for (i in slides.indices) {
            dots[i].setTextColor(if (i == currentPage) colorActive else colorInactive)
        }
    }

    private inner class TutorialViewPageAdapter : PagerAdapter() {
        override fun instantiateItem(container: ViewGroup, position: Int): Any {
            val layoutInflater = getSystemService(LAYOUT_INFLATER_SERVICE) as LayoutInflater
            val view = layoutInflater.inflate(slides[position], container, false)
            container.addView(view)
            return view
        }

        override fun getCount(): Int {
            return slides.size
        }

        override fun isViewFromObject(view: View, anObject: Any): Boolean {
            return view === anObject
        }

        override fun destroyItem(container: ViewGroup, position: Int, anObject: Any) {
            val view = anObject as View
            container.removeView(view)
        }
    }

    companion object {
        // Use Unicode character "Bullet" (decimal code 8226) as text.
        private const val DOT = 8226.toChar().toString()
    }
}
